####Data analysis and visualization 
###Seasonal impacts on gut microbial composition of the Eastern subterranean termites (Blattodea: Rhinotermitidae)
######R. Sapkota, C. H. Nakatsu and M. E. Scharf#####

##Install and load packages
if (!requireNamespace("BiocManager", quietly = TRUE))
   install.packages("BiocManager")
install.packages("ape")
install.packages("phangorn")


library (BiocManager)
library(phyloseq)
library(naniar)
library(zoo)
library(ape)
library(dplyr)
library(ggplot2)
library(gplots)
library(lme4)
library(phangorn)
library(phyloseq)
library(plotly)
library(tidyr)
library(vegan)

#Set directory and load Data
OTU = read.table("Transposed-rarified.txt", header=TRUE,check.names = FALSE, sep="\t", stringsAsFactors = F)
tax = read.table("taxonomy.tsv", header=TRUE, sep="\t")
meta = read.table("termite-meta.csv", header=TRUE, row.names=1, sep=",")
evenness = read.table("evenness.tsv", header=TRUE, row.names=1, sep="\t")
faith_pd = read.table("faith_pd.tsv", header=TRUE,row.names=1, sep="\t")
observed_features = read.table("observed_features.tsv", header=TRUE, row.names=1, sep="\t")
shannon = read.table("shannon.tsv", header=TRUE, row.names=1, sep="\t")

###OTU table
str(OTU)
row.names(OTU) = OTU[,1]
OTU.clean = OTU[,-1]
str(OTU.clean)
###Taxonomy table
library(tidyr)
library(devtools)
library(tidyselect)
library(naniar)
tax2 = separate(tax, Taxon, into = c("Domain", "Phylum", "Class", "Order", "Family", "Genus", "Species"), sep=";")
#All the strings that need to be removed and replaced with NA
na_strings <- c(" s__", " g__", " f__", " o__", " c__")
#str(tax.clean)
tax3 = replace_with_na_all(tax2, condition = ~.x %in% na_strings)

tax3[] <- t(apply(tax3, 1, zoo::na.locf))
tax3 = tax3[,-1]
tax3_mat <- as.matrix(tax3)

#Get the row names back
row.names(tax3_mat) <- tax2[,1]
tax.clean <- as.data.frame(tax3_mat)

#Remove all the OTUs that don't occur in our OTU.clean data set
tax.clean = tax.clean[row.names(tax.clean) %in% colnames(OTU.clean),]
##Alpha Diversity tables
meta = read.table("termite-meta.csv", header=TRUE, row.names=1, sep=",")
alpha_diversity = merge(faith_pd, evenness, by.x = 0, by.y = 0)
alpha_diversity = merge(alpha_diversity, observed_features, by.x = "Row.names", by.y = 0)
alpha_diversity = merge(alpha_diversity, shannon, by.x = "Row.names", by.y = 0)

meta = merge(meta, alpha_diversity, by.x = 0, by.y = "Row.names")
row.names(meta) = meta$Row.names
meta = meta[,-1]
##Order the data
OTU.clean = OTU.clean[order(row.names(OTU.clean)),]
meta = meta[order(row.names(meta)),]
set.seed(215)
##Explore alpha metrics
#Create 2x2 plot environment 
par(mfrow = c(2, 2))
#Plots
hist(meta$shannon, main="Shannon diversity", xlab="", breaks=10)
hist(meta$faith_pd, main="Faith phylogenetic diversity", xlab="", breaks=10)
hist(meta$pielou_evenness, main="Evenness", xlab="", breaks=15)
hist(meta$observed_features, main="Observed features", xlab="", breaks=15)
#To test for normalcy statistically, run the Shapiro-Wilk test of normality.
shapiro.test(meta$shannon)
shapiro.test(meta$pielou_evenness)
shapiro.test(meta$faith_pd)
shapiro.test(meta$observed_features)
**Non-normally distributed metrics**
meta$Months = factor(meta$Months, c("May","Jun", "Jul", "Aug", "Sep", "Oct"))
levels(meta$Months)
#USed the statistics from qiime2 results
kruskal.test(faith_pd ~ Months, data=meta)
kruskal.test(observed_features ~ Months, data=meta)
kruskal.test(shannon_entropy ~ Months, data=meta)
kruskal.test(pielou_evenness ~ Months, data=meta)
#Pairwise comparisons using Dunn test and "Benjamini-Hochberg" adjustment methods
library(lattice)
library(FSA)
dunnTest(observed_features ~ Months, data=meta, method="bh")
dunnTest(shannon_entropy ~ Months, data=meta, method="bh") 
dunnTest(pielou_evenness ~ Months, data=meta, method="bh") 

#Create plot environment
par(mfrow = c(1, 2))
#Plot
boxplot(pielou_evenness ~ Months, data=meta, ylab="Pielou evenness",col="white")
boxplot(faith_pd ~ Months, data=meta, ylab="Faith phylogenetic diversity",col="white")
par(mfrow = c(1, 2))
boxplot(shannon_entropy ~ Months, data=meta, ylab="Shannon diversity", col="white")
boxplot(observed_features ~ Months, data=meta, ylab="Observed features", col="white")

##Visualization
BC.nmds = metaMDS(OTU.clean, distance="bray", k=2, trymax=1000, binary= TRUE)
Jac.nmds = metaMDS(OTU.clean, distance="jaccard", k=2, trymax=1000, binary= TRUE)
#Solution reached around 20 iterations and stress is low (<0.1), 2-axis are sufficient to view the data.
my_colors = c("blue", "green", "red", "black", "purple", "magenta", "cyan")
par(mfrow = c(1, 2))
#Create a blank plot for the nmds
plot(BC.nmds, type="n", main="Bray-Curtis dissimilarity") 
#Add the points colored by age
points(BC.nmds$points, display="Months", pch=20, col=my_colors[meta$Months])
#Add a legend
legend(0.8, 1.0, legend=levels(meta$Months), col=my_colors, pch=20)

my_colors = c("blue", "green", "red", "black", "purple", "magenta", "cyan")
#par(mfrow = c(1, 1))
plot(Jac.nmds, type="n", main="Jaccard index") 
#Add the points colored by group
points(Jac.nmds$points, display="Months", pch=20, col=my_colors[meta$Months])
#Add a legend
legend(0.65, 1.2, legend=levels(meta$Months), col=my_colors, pch=20)


library(ggplot2)
BC.nmds$stress
nmds <-as.data.frame(BC.nmds$points)
metanmds <- merge(meta, nmds, by.x = 0, by.y = 0)
row.names(metanmds) <- metanmds[,1]
metanmds <- metanmds[,-1]
str(metanmds)
metanmds$Months <- factor(metanmds$Months)

NMDS.mean <- aggregate(metanmds, list(group=metanmds$Months), mean)
colnames(NMDS.mean) <- c('design', 'groupX', 'groupY')


metanmds <- merge(metanmds, NMDS.mean, by.x = "Months", by.y="design")

str(metanmds)

ggplot(metanmds, aes(x=MDS1, y=MDS2)) +
  geom_point(aes(color=Months, shape= Months)) +
  labs(x='NMDS 1', y= 'NMDS 2', caption = paste('Ordination stress: ', round(BC.nmds$stress, digits = 2))) +
  stat_ellipse(aes(color=Months), level = 0.95) +
  theme(legend.title = element_blank()) 

ggsave("nmds_ellipses-bray_all1.pdf", height = 5, width = 7)

#Jaccard Like bray-curtis
Jac.nmds$stress

Jnmds <-as.data.frame(Jac.nmds$points)
Jmetanmds <- merge(meta, Jnmds,by.x = 0, by.y = 0)
row.names(Jmetanmds) <- Jmetanmds[,1]
Jmetanmds <- Jmetanmds[,-1]
str(Jmetanmds)
Jmetanmds$Months <- factor(Jmetanmds$Months)

JNMDS.mean <- aggregate(Jmetanmds,list(group=Jmetanmds$Months), mean)
colnames(JNMDS.mean) <- c('design', 'groupX', 'groupY')
Jmetanmds <- merge(Jmetanmds, JNMDS.mean, by.x = "Months", by.y="design")
str(Jmetanmds)

ggplot(Jmetanmds, aes(x=MDS1, y=MDS2)) +
  geom_point(aes(color=Months, shape=Months)) +
  labs(x='NMDS 1', y= 'NMDS 2', caption = paste('Ordination stress: ', round(Jac.nmds$stress, digits = 2))) +
  stat_ellipse(aes(color=Months), level = 0.95) +
  theme(legend.title = element_blank()) 

ggsave("nmds_jac-ellipses_all.pdf", height = 3, width = 4)

###Phylogentic-based metrics
install.packages ("qiime2R")
if (!requireNamespace("devtools", quietly = TRUE)){install.packages("devtools")}
devtools::install_github("jbisanz/qiime2R") # current version is 0.99.20
install.packages("devtools")

library(qiime2R)
library (devtools)

#Weighted Unifrac-1-2
library(qiime2R)
library(dplyr)
metadata = read.table("termite-meta.csv", header=TRUE, sep=",")
wunifrac <- read_qza ("weighted_unifrac_pcoa_results.qza")

wunifrac$data$Vectors %>%
  select(SampleID, PC1, PC2) %>%
  left_join(metadata) %>%
  #left_join(shannon_diversity) %>%
  ggplot(aes(x=PC1, y=PC2, color=`Months`, shape=)) +
  geom_point(alpha=1) + #alpha controls transparency and helps when points are overlapping
  theme_q2r() +
  #scale_shape_manual(values=c(16,1), name="Months") + #see http://www.sthda.com/sthda/RDoc/figure/graphs/r-plot-pch-symbols-points-in-r.png for numeric shape codes
  scale_size_continuous(name="Shannon Diversity") +
  scale_color_discrete(name="Months")
  ggsave("wUF-PC1-2.pdf", height=3, width=3, device="pdf") # save a PDF 3 inches by 4 inches
  ggsave("wUF-PC1-2_bigger.png", height=3, width=4, device="png", dpi=300) # save a PDF 3 inches by 4 inches
  ggsave("wUF-PC1-2_bigger.pdf", height=3, width=4, device="pdf", dpi=300) # save a PDF 3 inches by 4 inches

#Weighted Unifrac-1-3
library(qiime2R)
metadata = read.table("termite-meta.csv", header=TRUE, sep=",")
wunifrac <- read_qza ("weighted_unifrac_pcoa_results.qza")

wunifrac$data$Vectors %>%
  select(SampleID, PC1, PC3) %>%
  left_join(metadata) %>%
  #left_join(shannon) %>%
  ggplot(aes(x=PC3, y=PC1, color=`Months`, shape=)) +
  geom_point(alpha=1) + #alpha controls transparency and helps when points are overlapping
  theme_q2r() +
  #scale_shape_manual(values=c(16,1), name="Months") + #see http://www.sthda.com/sthda/RDoc/figure/graphs/r-plot-pch-symbols-points-in-r.png for numeric shape codes
  scale_size_continuous(name="Shannon Diversity") +
  scale_color_discrete(name="Months")
  ggsave("wUF-PCoA1-3.pdf", height=3, width=4, device="pdf", dpi=300) # save a PDF 3 inches by 4 inches
  ggsave("wUF-PCoA_bigger_1,3.png", height=3, width=4, device="png", dpi=300) # save a PDF 3 inches by 4 inches

#Weighted Unifrac- 2-3
library(qiime2R)
metadata = read.table("termite-meta.csv", header=TRUE, sep=",")
wunifrac <- read_qza ("weighted_unifrac_pcoa_results.qza")

wunifrac$data$Vectors %>%
  select(SampleID, PC2, PC3) %>%
  left_join(metadata) %>%
  #left_join(shannon) %>%
  ggplot(aes(x=PC3, y=PC2, color=`Months`, shape=)) +
  geom_point(alpha=1) + #alpha controls transparency and helps when points are overlapping
  theme_q2r() +
  #scale_shape_manual(values=c(16,1), name="Months") + #see http://www.sthda.com/sthda/RDoc/figure/graphs/r-plot-pch-symbols-points-in-r.png for numeric shape codes
  scale_size_continuous(name="Shannon Diversity") +
  scale_color_discrete(name="Months")
  ggsave("wUF-PCoA_bigger_2-3.pdf", height=3, width=4, device="pdf", dpi=300) # save a PDF 3 inches by 4 inches
  ggsave("wUF-PCoA_bigger_2-3.png", height=3, width=4, device="png", dpi=300) # save a PDF 3 inches by 4 inches

#UnWeighted Unifrac-2-3
library(qiime2R)
metadata = read.table("termite-meta.csv", header=TRUE, sep=",")
unwunifrac <- read_qza ("unweighted_unifrac_pcoa_results.qza")

unwunifrac$data$Vectors %>%
  select(SampleID, PC3, PC2) %>%
  left_join(metadata) %>%
  #left_join(shannon) %>%
  ggplot(aes(x=PC3, y=PC2, color=`Months`, shape=)) +
  geom_point(alpha=1) + #alpha controls transparency and helps when points are overlapping
  theme_q2r() +
  #scale_shape_manual(values=c(16,1), name="Months") + #see http://www.sthda.com/sthda/RDoc/figure/graphs/r-plot-pch-symbols-points-in-r.png for numeric shape codes
  scale_size_continuous(name="Shannon Diversity") +
  scale_color_discrete(name="Months")
  ggsave("UnwUF-PCoA-2-3.pdf", height=3, width=3, device="pdf") # save a PDF 3 inches by 4 inches
  ggsave("UnwUF-PCoA_bigger-2-3.png", height=4, width=5, device="png") # save a PDF 3 inches by 4 inches
  ggsave("UnwUF-PCoA_bigger-2-3.pdf", height=3, width=4, device="pdf") # save a PDF 3 inches by 4 inches

#UnWeighted Unifrac-1-2
library(qiime2R)
metadata = read.table("termite-meta.csv", header=TRUE, sep=",")
unwunifrac <- read_qza ("unweighted_unifrac_pcoa_results.qza")

unwunifrac$data$Vectors %>%
  select(SampleID, PC1, PC2) %>%
  left_join(metadata) %>%
  #left_join(shannon) %>%
  ggplot(aes(x=PC1, y=PC2, color=`Months`, shape=)) +
  geom_point(alpha=1) + #alpha controls transparency and helps when points are overlapping
  theme_q2r() +
  #scale_shape_manual(values=c(16,1), name="Months") + #see http://www.sthda.com/sthda/RDoc/figure/graphs/r-plot-pch-symbols-points-in-r.png for numeric shape codes
  scale_size_continuous(name="Shannon Diversity") +
  scale_color_discrete(name="Months")
  ggsave("UnwUF-PCoA-1-2.pdf", height=3, width=3, device="pdf") # save a PDF 3 inches by 4 inches
  ggsave("UnwUF-PCoA_bigger-1-2.png", height=4, width=4, device="png") # save a PDF 3 inches by 4 inches
  ggsave("UnwUF-PCoA_bigger-1-2.pdf", height=3, width=4, device="pdf") # save a PDF 3 inches by 4 inches

#UnWeighted Unifrac-1-3
unwunifrac$data$Vectors %>%
  select(SampleID, PC1, PC3) %>%
  left_join(metadata) %>%
  #left_join(shannon) %>%
  ggplot(aes(x=PC1, y=PC3, color=`Months`, shape=)) +
  geom_point(alpha=1) + #alpha controls transparency and helps when points are overlapping
  #stat_ellipse(aes(color=`Months`), level = 0.95) +
  theme_q2r() +
  #scale_shape_manual(values=c(16,1), name="Months") + #see http://www.sthda.com/sthda/RDoc/figure/graphs/r-plot-pch-symbols-points-in-r.png for numeric shape codes
  scale_size_continuous(name="Shannon Diversity") +
  scale_color_discrete(name="Treatments")
  ggsave("unwUF-PCoA-ellipse-1-3.pdf", height=3, width=4, device="pdf", dpi=300) # save a PDF 3 inches by 4 inches
  ggsave("unwUF-PCoA-ellipse_bigger-1-3.pdf", height=4, width=5, device="pdf", dpi=300) # save a PDF 3 inches by 4 inches
##Bar charts visualizations

install.packages("ggplot2")
install.packages("reshape2")
library(ggplot2)
library(reshape2)
librarypackageVersion("phyloseq")
install.packages("phyloseq")
library("phyloseq")
library(dplyr)

##Metadata should be in .txt format to run
library(qiime2R)
physeq<-qza_to_phyloseq (
    features="dada2-final-table.qza",
    tree="rooted-tree.qza",
    "taxonomy.qza",
    metadata = "Ntermite-meta.txt"
    )
#Abundance
plot_bar(physeq, fill="Phylum")
plot_bar(physeq, x="Months", fill="Phylum") 
plot_bar(physeq, x="Months", fill="Phylum") + geom_bar(aes(color=Phylum, fill=Phylum), stat="identity", position="stack")

#And only showing the top 15 most abundant phyla
top15P.names = sort(tapply(taxa_sums(physeq), tax_table(physeq)[, "Phylum"], sum), TRUE)[1:15]
#Cut down the physeq data to only the top 5 Phyla
top15P = subset_taxa(physeq, Phylum %in% names(top15P.names))
#Plot
plot_bar(top15P, x="Months", fill="Phylum") + geom_bar(aes(color=Phylum, fill=Phylum), stat="identity", position="stack")

#Relative abundance at phylum level

my_colors <- c(
  '#a6cee3','#1f78b4','#b2df8a','#33a02c','#fb9a99','#e31a1c',
  '#fdbf6f','#ff7f00','#cab2d6','#6a3d9a','#ffff99','#b15928', 
  "#CBD588", "#5F7FC7", "orange","#DA5724", "#508578", "#CD9BCD",
   "#AD6F3B", "#673770","#D14285", "#652926", "#C84248", 
  "#8569D5", "#5E738F","#D1A33D", "#8A7C64", "#599861", "black"
)
my_level <- c("Phylum")
#rm(taxa.summary)
ml ="Phylum"
for(ml in my_level){
  print(ml)
  
    taxa.summary <- physeq %>%
    tax_glom(taxrank = ml) %>%                     
    transform_sample_counts(function(x) {x/sum(x)} ) %>% 
    psmelt()  %>%                                           
    group_by(Months, get(ml)) %>%
    summarise(Abundance.average=mean(Abundance)) 
  names(taxa.summary)[2] <- ml
  
  physeq.taxa.average <- taxa.summary %>% 
    group_by(get(ml)) %>%
    summarise(overall.average=mean(Abundance.average))
  names(physeq.taxa.average)[1] <- ml
  
 
  physeq_meta <- merge(taxa.summary, physeq.taxa.average)
    
  abund_filter <- 0.001
  physeq_meta_filtered <- filter(physeq_meta, overall.average>abund_filter)
  str(physeq_meta_filtered)
  
  physeq_meta_filtered$Months = factor(physeq_meta_filtered$Months, c("May", "Jun", "Jul", "Aug", "Sep", "Oct"))
  unique(physeq_meta_filtered$`get(ml)`)
  

#Plot-relative abundance-barplot@ phylum level
   ggplot(physeq_meta_filtered, aes(x = Months, y = Abundance.average, fill = Phylum)) + 
    #facet_grid(.~subject) +
    geom_bar(stat = "identity") +
    scale_fill_manual(values = my_colors) +
    # Remove x axis title
    #theme(axis.title.x = element_blank()) + 
    ylim(c(0,1)) +
    guides(fill = guide_legend(reverse = F, keywidth = .5, keyheight = .5, ncol = 1)) +
    theme(legend.text=element_text(size=8)) +
    #theme(legend.position="bottom") +
    theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5)) +
    ylab("Relative Abundance") + xlab("Months") +
    ggtitle(paste0(ml, " (>", abund_filter * 100,"%) Composition of microbiome samples")) 
ggsave(paste0("BarPlot_AllSamples_phylum0.0001.png"), height = 5, width= 6)
ggsave(paste0("BarPlot_AllSamples_phylum0.0001.pdf"), height = 5, width= 6)
}

#Relative abundance at family level
my_level <- c("Family")

my_colors <- c(
  '#a6cee3','#1f78b4','#b2df8a','#33a02c','#fb9a99','#e31a1c',
  '#fdbf6f','#ff7f00','#cab2d6','#6a3d9a','#ffff99','#b15928', 
  "#CBD588", "#5F7FC7", "orange","#DA5724", "#508578", "#CD9BCD",
   "#AD6F3B", "#673770","#D14285", "#652926", "#C84248", 
  "#8569D5", "#5E738F","#D1A33D", "#8A7C64", "#599861", "black"
)
ml ="Family"
for(ml in my_level){
  print(ml)
  
  taxa.summary.fam <- physeq %>%
    tax_glom(taxrank = ml) %>%                     
    transform_sample_counts(function(x) {x/sum(x)} ) %>% 
    psmelt()  %>%                                           
    group_by(Months, get(ml)) %>%
    summarise(Abundance.average=mean(Abundance)) 
  names(taxa.summary.fam)[2] <- ml
  
  physeq.taxa.average.fam <- taxa.summary.fam %>% 
    group_by(get(ml)) %>%
    summarise(overall.average=mean(Abundance.average))
  names(physeq.taxa.average.fam)[1] <- ml
  
  physeq_meta.fam <- merge(taxa.summary.fam, physeq.taxa.average.fam)
    
  abund_filter <- 0.01
  physeq_meta_filtered.fam <- filter(physeq_meta.fam, overall.average>abund_filter)
  str(physeq_meta_filtered.fam)
  
   physeq_meta_filtered.fam$Months = factor(physeq_meta_filtered.fam$Months, c("May", "Jun", "Jul", "Aug", "Sep", "Oct"))
  unique(physeq_meta_filtered.fam$`get(ml)`)
  
#Plot-relative abundance-barplot@ family level
  ggplot(physeq_meta_filtered.fam, aes(x = Months, y = Abundance.average, fill = Family)) + 
    #facet_grid(.~subject) +
    geom_bar(stat = "identity", position = "fill") +
    scale_fill_manual(values = my_colors) +
    # Remove x axis title
    #theme(axis.title.x = element_blank()) + 
    ylim(c(0,1)) +
    guides(fill = guide_legend(reverse = F, keywidth = .5, keyheight = .5, ncol = 1)) +
    theme(legend.text=element_text(size=8)) +
    #theme(legend.position="bottom") +
    theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5)) +
    ylab("Relative Abundance") + xlab("Months") +
    ggtitle(paste0(ml, " (>", abund_filter * 100,"%) Composition of microbiome samples")) 
ggsave(paste0( "BarPlot_AllSamples.fam.png"), height = 5, width= 6)
ggsave(paste0( "BarPlot_AllSamples.fam.pdf"), height = 5, width= 6)
}


# Relative abundance at genus level
my_level <- c("Genus")
my_colors <- c(
  '#a6cee3','#1f78b4','#b2df8a','#33a02c','#fb9a99','#e31a1c',
  '#fdbf6f','#ff7f00','#cab2d6','#6a3d9a','#ffff99','#b15928', 
  "#CBD588", "#5F7FC7", "orange","#DA5724", "#508578", "#CD9BCD",
   "#AD6F3B", "#673770","#D14285", "#652926", "#C84248", 
  "#8569D5", "#5E738F","#D1A33D", "#8A7C64", "#599861", "black", "powderblue","purple","purple1","purple2",   "purple3","sienna2", "palevioletred2", "steelblue1", "tan1", "tomato1", "turquoise1", "violetred", "whitesmoke", "yellow4", "aquamarine3", "azure3", "bisque2", "blueviolet", "brown4", "burlywood4", "cadetblue4", "chocolate4", "deeppink3", "firebrick1", "forestgreen", "gold2", "gray1",  '#a6cee3','#1f78b4','#b2df8a','#33a02c','#fb9a99','#e31a1c',
  '#fdbf6f','#ff7f00','#cab2d6','#6a3d9a','#ffff99','#b15928', 
  "#CBD588", "#5F7FC7", "orange","#DA5724", "#508578", "#CD9BCD",
   "#AD6F3B", "#673770","#D14285", "#652926", "#C84248", 
  "#8569D5", "#5E738F","#D1A33D", "#8A7C64", "#599861", "black")
ml ="Genus"
for(ml in my_level){
  print(ml)
  
  taxa.summary.gen <- physeq %>%
    tax_glom(taxrank = ml) %>%                     
    transform_sample_counts(function(x) {x/sum(x)} ) %>% 
    psmelt()  %>%                                           
    group_by(Months, get(ml)) %>%
    summarise(Abundance.average=mean(Abundance)) 
  names(taxa.summary.gen)[2] <- ml
  
  physeq.taxa.average.gen <- taxa.summary.gen %>% 
    group_by(get(ml)) %>%
    summarise(overall.average=mean(Abundance.average))
  names(physeq.taxa.average.gen)[1] <- ml
  
    physeq_meta.gen <- merge(taxa.summary.gen, physeq.taxa.average.gen)

  abund_filter <- 0.001
  physeq_meta_filtered.gen <- filter(physeq_meta.gen, overall.average>abund_filter)
  str(physeq_meta_filtered.gen)
  
   physeq_meta_filtered.gen$Months = factor(physeq_meta_filtered.gen$Months, c("May", "Jun", "Jul", "Aug", "Sep", "Oct"))
  unique(physeq_meta_filtered.gen$`get(ml)`)
  
#Plot-relative abundance-barplot@ Genus level 
   ggplot(physeq_meta_filtered.gen, aes(x = Months, y = Abundance.average, fill = Genus)) + 
    #facet_grid(.~subject) +
    geom_bar(stat = "identity", position = "fill") +
    scale_fill_manual(values = my_colors) +
    # Remove x axis title
    #theme(axis.title.x = element_blank()) + 
    ylim(c(0,1)) +
    guides(fill = guide_legend(reverse = F, keywidth = .5, keyheight = .5, ncol = 1)) +
    theme(legend.text=element_text(size=8)) +
        #theme(legend.position="bottom") +
    theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5)) +
    ylab("Relative Abundance") + xlab("Months") +
    ggtitle(paste0(ml, " (>", abund_filter * 100,"%) Composition of microbiome samples")) 
ggsave(paste0( "BarPlot_0.1.gen.png"), height = 5, width= 6)
ggsave(paste0( "BarPlot_0.1.gen.pdf"), height = 5, width= 6)

#Spearman correlation between relative abundances of bacterial taxa and average temperatures.
#With average collection-day-temperature
library(ggplot2)
data1=read.csv(file="Collectionday-temp-taxa.csv",header=TRUE, sep = ",", row.names = NULL)
cormat1=cor(data1, method = "spearman")
cormat1

### Determining significance level
install.packages("Hmisc")
library(Hmisc)
cormat2 <- rcorr(as.matrix(data1), type = "spearman")
cormat2
# Extract the correlation coefficients
cormat=cormat2$r
# Extract p-values
pmat=cormat2$P
##### Organizing the results/values
++++++++++++++++++++++++++++
  # flattenCorrMatrix
  # ++++++++++++++++++++++++++++
  # cormat : matrix of the correlation coefficients
  # pmat : matrix of the correlation p-values
  
   flattenCorrMatrix <- function(cormat, pmat) {
    ut <- upper.tri(cormat)
    data.frame(
      row = rownames(cormat)[row(cormat)[ut]],
      column = rownames(cormat)[col(cormat)[ut]],
      cor  =(cormat)[ut],
      p = pmat[ut]
    )
  }

Results1=flattenCorrMatrix(cormat2$r, cormat2$P)
write.csv(Results1,"Same-day-Results2.csv")

#With average monthly-temperature
library(ggplot2)
data2=read.csv(file="monthly-temp-taxa.csv",header=TRUE, sep = ",", row.names = NULL)
cormat3=cor(data1, method = "spearman")
cormat3

### Determining significance level
install.packages("Hmisc")
library(Hmisc)
cormat4 <- rcorr(as.matrix(data2), type = "spearman")
cormat4
# Extract the correlation coefficients
cormat5=cormat4$r
# Extract p-values
pmat1=cormat4$P
##### Organizing the results/values
++++++++++++++++++++++++++++
  # flattenCorrMatrix
  # ++++++++++++++++++++++++++++
  # cormat : matrix of the correlation coefficients
  # pmat : matrix of the correlation p-values
  
   flattenCorrMatrix <- function(cormat5, pmat1) {
    ut <- upper.tri(cormat5)
    data.frame(
      row = rownames(cormat5)[row(cormat5)[ut]],
      column = rownames(cormat5)[col(cormat5)[ut]],
      cor  =(cormat5)[ut],
      p = pmat1[ut]
    )
  }

Results2=flattenCorrMatrix(cormat4$r, cormat4$P)
write.csv(Results2,"Av-monthly-Results2.csv")

#With average week's temperature prior to collection day
library(ggplot2)
data1=read.csv(file="Week-av-temp-taxa.csv",header=TRUE, sep = ",", row.names = NULL)
cormat1=cor(data1, method = "spearman")
cormat1

### Determining significance level
install.packages("Hmisc")
library(Hmisc)
cormat2 <- rcorr(as.matrix(data1), type = "spearman")
cormat2
# Extract the correlation coefficients
cormat=cormat2$r
# Extract p-values
pmat=cormat2$P
##### Organizing the results/values
++++++++++++++++++++++++++++
  # flattenCorrMatrix
  # ++++++++++++++++++++++++++++
  # cormat : matrix of the correlation coefficients
  # pmat : matrix of the correlation p-values
  
   flattenCorrMatrix <- function(cormat, pmat) {
    ut <- upper.tri(cormat)
    data.frame(
      row = rownames(cormat)[row(cormat)[ut]],
      column = rownames(cormat)[col(cormat)[ut]],
      cor  =(cormat)[ut],
      p = pmat[ut]
    )
  }

Results1=flattenCorrMatrix(cormat2$r, cormat2$P)
write.csv(Results1,"Wk.av-Results2.csv")

#Spearman-correlation-temperature with alpha diversity matrices

library(ggplot2)

data1=read.csv(file="Alpha-temp-corr.csv",header=TRUE, sep = ",", row.names = NULL)
cormat1=cor(data1, method = "spearman")
cormat1

### Determining significance level
install.packages("Hmisc")
library(Hmisc)
cormat2 <- rcorr(as.matrix(data1), type = "spearman")
cormat2
# Extract the correlation coefficients
cormat=cormat2$r
# Extract p-values
pmat=cormat2$P
##### Organizing the results/values
++++++++++++++++++++++++++++
  # flattenCorrMatrix
  # ++++++++++++++++++++++++++++
  # cormat : matrix of the correlation coefficients
  # pmat : matrix of the correlation p-values
  
   flattenCorrMatrix <- function(cormat, pmat) {
    ut <- upper.tri(cormat)
    data.frame(
      row = rownames(cormat)[row(cormat)[ut]],
      column = rownames(cormat)[col(cormat)[ut]],
      cor  =(cormat)[ut],
      p = pmat[ut]
    )
  }

Results1=flattenCorrMatrix(cormat2$r, cormat2$P)
write.csv(Results1,"alpha-temp-corr-Results2.csv")
#No significant interaction between temperature and alpha diversity metrics was observed.
